core/src/docs/rfcs/2602_object_versioning.md

```diff
@@ -0,0 +1,144 @@
+But here it is imperceptible to the user.
+
+## What operations are supported with object versioning?
+
+When object versioning is enabled, the following operations will be supported:
+
+- `stat`: Get the metadata of an object with specific version ID.
+- `read`: Read a specific version of an object.
+- `delete`: Delete a specific version of an object.
+
+Code example:
+
+```rust
+// To get the current version ID of a file
+let meta = op.stat("path/to/file").await?;
+let version_id = meta.version().expect("just for example");
+
+// To fetch the metadata of specific version of a file
+let meta = op.stat_with("path/to/file").version("version_id").await?;
+let version_id = meta.version().expect("just for example"); // get the version ID
+
+// To read an file with specific version ID
+let content = op.read_with("path/to/file").version("version_id").await?;
+
+// To delete an file with specific version ID
+op.delete_with("path/to/file").version("version_id").await?;
+```
+
+## What are the benefits of object versioning?
+
+With object versioning, users can:
+
+- Track the history of a file.
+- Implement optimistic concurrency control.
+- Implement a simple backup system.
+
+# Reference-level explanation
+
+Those operations with object version are different from the normal operations:
+
+- `stat`: when getting the metadata of a file, it will always get the metadata of the latest version of the file if no version ID is specified. And there will be a new field `version` in the metadata to indicate the version ID of the file.
+- `read`: when reading a file, it will always read the latest version of the file if no version ID is specified.
+- `delete`: when deleting a file, it will always delete the latest version of the file if no version ID is specified. And users will not be able to read this file without specifying the version ID, unless they specify a version not be deleted.
+
+And with object versioning, when writing an object, 
+it will always create a new version of the object than overwrite the old version. 
+But here it is imperceptible to the user. 
+Because the version id is generated by the service itself, it cannot be specified by the user and user cannot override the historical version.
+
+To implement object versioning, we will do the following:
+
+- Add a new field `version` to `OpStat`, `OpRead` and `OpDelete` struct.
+- Add a new field `version` to `ObjectMetadata` struct.
+- Add a new property(setter) `version` to the return value of `stat_with`, `read_with` method.
+- Add a new method `delete_with` and add a new property(setter) `version` to the return value of `delete_with` method.
+
+For service backend, it should support the following operations:
+
+- `stat`: Get the metadata of an object with specific version ID.
+- `read`: Read a specific version of an object.
+- `delete`: Delete a specific version of an object.
+
+## reference:
```

Comment:
Please don't add `:` in title, and reference should be moved to `Rationale and alternatives`